src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -461,96 +556,106 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
+                    // before we have a chance to process cancellation and tear everything down. Just ignore this.
+                    if (_resetException != null)
                     {
                         return;
                     }
 
-                    // We should not retry request which have started being processed since the behavior might be unpredictable
-                    // I.e. some action might have been taken based on the received data.
-                    // We will bubble the exception and let user decide what to do.
-                    bool isRetriable = _state == StreamState.ExpectingStatus || _state == StreamState.ExpectingHeaders;
-                    Interlocked.CompareExchange(ref _abortException, abortException, null);
-                    _state = StreamState.Aborted;
-                    _canRetry = canRetry && isRetriable;
+                    // If the server told us the request has not been processed (via Last-Stream-ID on GOAWAY),
+                    // but we've already received some response data from the server, then the server lied to us.
+                    // In this case, don't allow the request to be retried.
+                    if (canRetry && _responseProtocolState != ResponseProtocolState.ExpectingStatus)
+                    {
+                        canRetry = false;
+                    }
 
-                    signalWaiter = _hasWaiter;
-                    _hasWaiter = false;
+                    _resetException = resetException;
+                    _canRetry = canRetry;
                 }
 
-                if (signalWaiter)
-                {
-                    _waitSource.SetResult(true);
-                }
+                CancelRequestBody();
+                CancelResponseBody();
             }
 
-            private void CheckIfDisposedOrAborted()
+            private void CheckResponseBodyState()
             {
                 Debug.Assert(Monitor.IsEntered(SyncObject));
 
-                if (_disposed)
-                {
-                    throw new ObjectDisposedException(nameof(Http2Stream));
-                }
-
-                if (_state == StreamState.Aborted)
+                if (_resetException != null)
                 {
                     if (_canRetry)
                     {
-                        throw new HttpRequestException(SR.net_http_request_aborted, _abortException, allowRetry: true);
+                        throw new HttpRequestException(SR.net_http_request_aborted, _resetException, allowRetry: true);
                     }
 
-                    throw new IOException(SR.net_http_request_aborted, _abortException);
+                    throw new IOException(SR.net_http_request_aborted, _resetException);
                 }
+
+                if (_responseCompletionState == StreamCompletionState.Failed)
+                {
+                    throw new IOException(SR.net_http_request_aborted);
```

Comment:
There are no additional details we can include about why it failed?

Reply:
We could. It gets a little weird trying to figure out what exception to propagate in some of these cases. But we probably could and should propagate something more useful. Let me look at it.